src/aof.zig

```diff
@@ -242,43 +247,61 @@ pub const AOF = struct {
     }
 
     pub fn close(self: *AOF) void {
-        std.posix.close(self.fd);
+        self.file.close();
     }
 
-    /// Write a message to disk. Once this function returns, the data passed in
-    /// is guaranteed to have been written using O_DIRECT and O_SYNC and
-    /// can be considered safely persisted for recovery purposes once this
-    /// call returns.
-    ///
-    /// We purposefully use standard disk IO here, and not IO_uring. It'll
-    /// be slower and have syscall overhead, but it's considerably more
-    /// battle tested.
-    ///
-    /// We don't bother returning a count of how much we wrote. Not being
-    /// able to fully write the entire payload is an error, not an expected
-    /// condition.
-    pub fn write(self: *AOF, message: *const Message.Prepare, options: struct {
-        replica: u64,
-        primary: u64,
-    }) !void {
+    /// Write a message to disk, with standard blocking IO but using the OS's page cache. The AOF
+    /// borrows durability from the write ahead log: if the AOF hasn't been flushed, and the machine
+    /// loses power, the op is guaranteed to still be in the WAL.
+    pub fn write(self: *AOF, message: *const Message.Prepare) !void {
+        assert(self.state == .writing);
+        assert(self.state.writing.unflushed < constants.journal_slot_count);
+
         var entry: AOFEntry align(constants.sector_size) = undefined;
         entry.from_message(
             message,
-            .{ .replica = options.replica, .primary = options.primary },
             &self.last_checksum,
         );
 
         const disk_size = entry.calculate_disk_size();
 
         const bytes = std.mem.asBytes(&entry);
+        try self.file.writeAll(bytes[0..disk_size]);
```

Comment:
Yeah, write here was bothering me _a lot_, thanks!

Reply:
Yeah, I think my original logic was very wrong, but it didn't manifest because it was also using direct IO!